File: node.go
package main

import (
	"fmt"
	"time"
	"runtime"
	
	// Библиотеки для работы функции node
	"crypto/sha1"
	"encoding/hex"
	"strconv"
)

// Функция аналогична функции node1 написанной на Ruby
func node1 (text string) int64 {
	h := sha1.New()
	h.Write([]byte(text))
	sha1_hash := hex.EncodeToString(h.Sum(nil))
	
	sha1_hash_len := len(sha1_hash)
	last := sha1_hash[sha1_hash_len-1:sha1_hash_len]
	
	value, _ := strconv.ParseInt(last, 16, 64) //int64
	return value % 8
}

// Этот вариант примерно на 30% быстрее
// Благодаря отказу от лишнего преобразования из одной системы счисления в другую
func node2 (text string) int64 {
	h := sha1.New()
	h.Write([]byte(text))
	mas := h.Sum(nil) // "hello world" -> [42 174 108 53 201 79 207 180 21 219 233 95 64 139 156 233 30 232 70 237]
	return int64(mas[len(mas)-1]) % 8 // Берем последний элемент массива. Это целое десятичное число. И считаем остаток от деления на 8
}

func main() {
	// Используем только 1 ядро процессора. Ruby выполняется на одном ядре. Не будем давать преимущество Go
	runtime.GOMAXPROCS(1)
	
	var (
		date_from int64
		date_to int64
		delta int64
	)
	
	// -------- node v1 -----------
	date_from = time.Now().UnixNano()
	
	for i := 1; i <= 1000000; i++ {
		node1("user:"+string(i))	
	}
	
	date_to = time.Now().UnixNano()
	
	delta = date_to - date_from
	
	fmt.Println("node1: ", delta)
	
	// -------- node v2 -----------
	date_from = time.Now().UnixNano()
	
	for i := 1; i <= 1000000; i++ {
		node2("user:"+string(i))	
	}
	
	date_to = time.Now().UnixNano()
	
	delta = date_to - date_from
	
	fmt.Println("node2: ", delta)
}

File: node.rb
require 'rubygems'
require 'digest/sha1'

# Стандартный вариант
def node0(text)
	return Integer('0x'+Digest::SHA1.hexdigest(text))%8
end

# Именно этот вариант сравниваем с node1 на Go.
def node1(text)
	s = Digest::SHA1.hexdigest(text)
	return Integer('0x'+s[s.size-1, 1]) % 8
end

# Этот вариант быстрее еще на 5% по сравнению с node1 на Ruby
def node11(text)
	s = Digest::SHA1.hexdigest(text)
	return ('0x'+s[s.size-1, 1]).hex % 8
end

# Самый быстрый вариант
def node2(text)
	return Digest::SHA1.digest(text)[-1].ord & 0x7
end

# Чуть-чуть уступает варианту node2
def node3(text)
	return Digest::SHA1.digest(text)[-1].ord % 8 
end

# Тестирование скорости работы
# ---------- node v0 -----------------
date_from = Time.now.to_f
for i in 1..1000000 do 
	node0("user:"+i.to_s)
end
date_to = Time.now.to_f

delta = date_to - date_from

puts 'node0:'+delta.to_s

# ---------- node v1 -----------------
date_from = Time.now.to_f
for i in 1..1000000 do 
	node1("user:"+i.to_s)
end
date_to = Time.now.to_f

delta = date_to - date_from

puts 'node1:'+delta.to_s

# ---------- node v11 -----------------
date_from = Time.now.to_f
for i in 1..1000000 do 
	node11("user:"+i.to_s)
end
date_to = Time.now.to_f

delta = date_to - date_from

puts 'node11:'+delta.to_s

# ---------- node v2 -----------------
date_from = Time.now.to_f
for i in 1..1000000 do 
	node2("user:"+i.to_s)
end
date_to = Time.now.to_f

delta = date_to - date_from

puts 'node2:'+delta.to_s

# ---------- node v3 -----------------
date_from = Time.now.to_f
for i in 1..1000000 do 
	node3("user:"+i.to_s)
end
date_to = Time.now.to_f

delta = date_to - date_from

puts 'node3:'+delta.to_s